### DISCUSSION TITLE
SDK: Can't find cuda_runtime_api.h

### DISCUSSION THREAD

**Discussion by Achder** [OE4T/meta-tegra#906] [Q&A] [answered]
I was tasked with compiling an existing project with our jetson nano devkit yocto SDK.
But the compiler can't seem to find the cuda_runtime_api.h

I have added the following to my image:

```
TOOLCHAIN_HOST_TASK_append = " nativesdk-cmake nativesdk-cmake-dev nativesdk-packagegroup-cuda-sdk-host"
IMAGE_INSTALL_append = " cuda-toolkit cuda-samples cudnn tensorrt-core tensorrt-samples"
```

I've searched the build dir on my build-machine and found:

```
./tmp/sysroots-components/x86_64/cuda-cudart-native/usr/local/cuda-10.2/include/cuda_runtime_api.h
./tmp/sysroots-components/armv8a_tegra/cuda-cudart/usr/local/cuda-10.2/include/cuda_runtime_api.h
```

But the installed SDK only includes:

```
./0.1/sysroots/aarch64-poky-linux/usr/local/cuda-10.2/include/cuda_runtime_api.h
```

I source the SDK as usual.
Inside the project folder I run:

```
cmake . -DCMAKE_TOOLCHAIN_FILE=$OE_MAKE_TOOLCHAIN_FILE
make
```

Results in: 

```
fatal error: cuda_runtime_api.h: No such file or directory
```

The cuda flags from cuda.sh get sourced correctly. nvcc and so on can be found correctly.
I'm guessing the header file is missing on the host side.
Any ideas?

**Comment by madisongh** [comment, marked as answer]
My SDKs are the same - target headers only in the target side of the SDK, and I have some CMake-based CUDA projects that build OK. Are you using CMake CUDA language support, FindCUDA, or something else?  Is there anything in your CMake files that might be overriding the compiler flags?  There should be a `-I` flag being passed in that points at the target-side CUDA tree where it's needed.

**Reply by Achder** [reply]
Do you have an example cmake file for a CUDA project I could look at? 
I'm pretty sure I need to adjust ours.

**Reply by madisongh** [reply]
There's a fairly simple example [here](https://github.com/madisongh/gpu-burn/blob/master/CMakeLists.txt).

**Reply by Achder** [reply]
Fixed the cuda_runtime_api.h issue by adding

```
set(MYSYSROOT $OECORE_TARGET_SYSROOT)
set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} --sysroot=${MYSYSROOT}" CACHE INTERNAL "" FORCE)
set(CMAKE_C_LINK_FLAGS "${CMAKE_C_LINK_FLAGS} --sysroot=${MYSYSROOT}" CACHE INTERNAL "" FORCE)
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} --sysroot=${MYSYSROOT}" CACHE INTERNAL "" FORCE)
set(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} --sysroot=${MYSYSROOT}" CACHE INTERNAL "" FORCE)
```

to my cmake.
But now it can't find <string> which is odd.

**Reply by madisongh** [reply]
Can't find what?

**Reply by Achder** [reply]
Oh whoops. 
Now it can‘t find a regular `#include <string>`

**Reply by madisongh** [reply]
Are your CMake files directly setting the flags?  Between the SDK environment setup script and the toolchain files for cmake included in the SDK, as long as you use the SDK-resident cmake binary to generate your makefiles, all settings for compilers, flags, etc. should be automatic and default to what's correct for the cross-compilation setup in the SDK.

**Reply by Achder** [reply]
For completeness:
Sorry, my bad. It was a bad CMake config. I rewrote a lot of it and now it compiles and links. Basically I forgot a $ENV{...} here and there...
